File: src/parser/typecasts.ts
import { Parser, match, noop, oneOf, seq, seqConst, _ } from '../typed-parser'
import { Expression } from '../ast'
import {
  expectIdentifier,
  identifier,
  sepExpectIdentifiers,
  stringConstant,
  symbol,
} from './token'
import { optional, parenthesized } from './utils'

const optionalPrecision: Parser<string> = seq(
  optional(parenthesized(match('[0-9]+')))
)((arg) => (arg === null ? '' : `(${arg})`))

const optionalDecimalPrecision: Parser<string> = seq(
  optional(
    parenthesized(
      seq(
        match('[0-9]+'),
        symbol(','),
        match('[0-9]+')
      )((p, _, s) => `${p}, ${s}`)
    )
  )
)((arg) => (arg === null ? '' : `(${arg})`))

const join = (...args: string[]): string => args.join(' ').trim()

const intervalFieldNames = [
  'YEAR TO MONTH',
  'DAY TO HOUR',
  'DAY TO MINUTE',
  'DAY TO SECOND',
  'HOUR TO MINUTE',
  'HOUR TO SECOND',
  'MINUTE TO SECOND',
  'YEAR',
  'MONTH',
  'DAY',
  'HOUR',
  'MINUTE',
  'SECOND',
]

const optionalIntervalFields: Parser<string> = seq(
  optional(
    oneOf(...intervalFieldNames.map((field) => sepExpectIdentifiers(field)))
  )
)((arg) => (arg === null ? '' : arg))

const optionalTimeZoneModifier: Parser<string> = oneOf(
  seq(sepExpectIdentifiers('with time zone'))(join),
  seq(optional(seq(sepExpectIdentifiers('without time zone'))(join)))(
    (arg) => arg || ''
  )
)

const specialTypeCastTargetType = (syntax: 'prefix' | 'psql'): Parser<string> =>
  oneOf(
    seq(sepExpectIdentifiers('bit varying'), optionalPrecision)(join),
    seq(sepExpectIdentifiers('bit'), optionalPrecision)(join),
    seq(sepExpectIdentifiers('character varying'), optionalPrecision)(join),
    sepExpectIdentifiers('double precision'),
    seq(
      oneOf(expectIdentifier('numeric'), expectIdentifier('decimal')),
      optionalDecimalPrecision
    )(join),
    seq(
      oneOf(expectIdentifier('time'), expectIdentifier('timestamp')),
      optionalPrecision,
      optionalTimeZoneModifier
    )(join),
    seq(
      expectIdentifier('interval'),
      // prefix typecast syntax doesn't support interval field names
      syntax === 'prefix' ? seqConst('', noop) : optionalIntervalFields,
      optionalPrecision
    )(join)
  )

export const psqlTypeCast: Parser<string> = seq(
  symbol('::'),
  oneOf(specialTypeCastTargetType('psql'), identifier),
  oneOf<'[]' | ''>(seqConst('[]', symbol('['), symbol(']')), seqConst('', _))
)((_, id, arraySuffix) => id + arraySuffix)

/**
 * Typecasts of the form `type 'string'`
 *
 * Example: TIMEZONE (4) WITH TIME ZONE '2020-02-02T12:34:56.789123'
 *
 */
export const prefixTypeCast: Parser<Expression.TypeCast> = seq(
  oneOf(specialTypeCastTargetType('prefix'), identifier),
  stringConstant
)((targetType: string, value: string) =>
  Expression.createTypeCast(Expression.createConstant(value), targetType)
)

File: src-rs/parser/typecasts.rs
use super::result::Result;
use crate::ast;
use crate::parser::keyword::Keyword;
use crate::parser::token::{identifier, keyword, keywords, string, symbol, __};
use crate::parser::utils::{parenthesized, prefixed, prefixed_, seq};
use nom::branch::alt;
use nom::character::complete::digit1;
use nom::combinator::{opt, recognize, value};
use nom::sequence::{preceded, terminated, tuple};
use nom::Parser;

fn optional_precision(input: &str) -> Result<()> {
    value((), opt(parenthesized(digit1)))(input)
}

fn optional_decimal_precision(input: &str) -> Result<()> {
    value((), opt(parenthesized(tuple((digit1, symbol(","), digit1)))))(input)
}

fn optional_interval_fields(input: &str) -> Result<()> {
    value(
        (),
        opt(alt((
            keywords(&[Keyword::Year, Keyword::To, Keyword::Month]),
            keywords(&[Keyword::Day, Keyword::To, Keyword::Hour]),
            keywords(&[Keyword::Day, Keyword::To, Keyword::Minute]),
            keywords(&[Keyword::Day, Keyword::To, Keyword::Second]),
            keywords(&[Keyword::Minute, Keyword::To, Keyword::Second]),
            keyword(Keyword::Year),
            keyword(Keyword::Month),
            keyword(Keyword::Day),
            keyword(Keyword::Hour),
            keyword(Keyword::Minute),
            keyword(Keyword::Second),
        ))),
    )(input)
}

fn optional_timezone_modifier(input: &str) -> Result<()> {
    value(
        (),
        opt(alt((
            keywords(&[Keyword::With, Keyword::Time, Keyword::Zone]),
            keywords(&[Keyword::Without, Keyword::Time, Keyword::Zone]),
        ))),
    )(input)
}

#[derive(Clone, Copy)]
enum Syntax {
    Psql,
    Prefix,
}

fn special_type_cast_target_type<'a>(syntax: Syntax) -> impl FnMut(&'a str) -> Result<'a, ()> {
    move |input: &str| {
        alt((
            prefixed_(&[Keyword::Bit, Keyword::Varying], optional_precision),
            prefixed(Keyword::Bit, optional_precision),
            prefixed_(&[Keyword::Character, Keyword::Varying], optional_precision),
            keywords(&[Keyword::Double, Keyword::Precision]),
            value(
                (),
                tuple((
                    alt((keyword(Keyword::Numeric), keyword(Keyword::Decimal))),
                    optional_decimal_precision,
                )),
            ),
            value(
                (),
                tuple((
                    alt((keyword(Keyword::Time), keyword(Keyword::Timestamp))),
                    optional_precision,
                    optional_timezone_modifier,
                )),
            ),
            move |input: &'a str| -> Result<()> {
                let (input, _) = keyword(Keyword::Interval).parse(input)?;
                let (input, _) = match syntax {
                    Syntax::Psql => optional_interval_fields.parse(input)?,
                    _ => (input, ()),
                };
                optional_precision.parse(input)
            },
        ))(input)
    }
}

pub fn psql_type_cast(input: &str) -> Result<&str> {
    terminated(
        preceded(
            symbol("::"),
            recognize(tuple((
                alt((
                    special_type_cast_target_type(Syntax::Psql),
                    value((), identifier),
                )),
                opt(value("[]", tuple((symbol("["), symbol("]"))))),
            ))),
        ),
        __,
    )(input)
}

/**
 * Typecasts of the form `type 'string'`
 *
 * Example: TIMEZONE (4) WITH TIME ZONE '2020-02-02T12:34:56.789123'
 */
pub fn prefix_typecast(input: &str) -> Result<ast::Expression> {
    seq(
        (
            alt((
                recognize(special_type_cast_target_type(Syntax::Prefix)),
                identifier,
            )),
            string,
        ),
        |(target_type, value)| ast::Expression::TypeCast {
            lhs: Box::new(ast::Expression::Constant(ast::Constant::String(value))),
            target_type,
        },
    )(input)
}
